README.md
# gncd-mail service #

My qmail-based email service, targeted at automatic messages + ezmlm
mailing lists.

Explicitly not targeted at end-user email.


## Existing branches ##

 * `master`: contains our final distribution, ready to rsync to a server
   and `./setup`
 * `src/netqmail`: the source of the netqmail project - we import the
   latest version available;
 * `patches/*`: all the patches that we track, either already merged
   into our distribution, or under evaluation for future integration;
 * `qmail/current`: main development branch for our qmail distribution -
   includes all patches we merged already.


ezmlm/ChangeLog
commit 2c81e75a7eccf84a46cea477622a5e1c432e0e6a
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Thu Nov 18 20:43:12 2010 -0600

    Tag release with today's date

commit a4c875df796885f693ec755c7dc5f1a8541d265d
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Thu Nov 18 20:43:31 2010 -0600

    Add TODO note

commit a525eb3719570b1bfb39ffd13cfe44b96d5b451b
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Thu Nov 18 20:22:04 2010 -0600

    tools/checklangs: Rewrote to produce more useful messages

commit faab89e250072f8eefbc773cf2d431f6afe3c52e
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Mon Nov 15 10:21:53 2010 -0600

    Add most -Wextra warnings, and fix nuisance warning in substdo.c

commit 8dd9da3f4136e95abbfa2f971605bcbcd0eca6ff
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Mon Nov 15 09:03:47 2010 -0600

    Remove unreachable returns

commit bd2c4b86b49c5ab37544752883c5933bcd4f3055
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Sun Nov 14 21:16:37 2010 -0600

    Change return type of main to int to allow compiling with clang

commit d80f13023cfe63c1fb53831b074aaf5dab45508f
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Sun Nov 14 21:10:05 2010 -0600

    ezmlm-dispatch: Remove unreachable return

commit a54a4f21c61e766638fa0314d7180d8a9fe5760e
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Fri Nov 12 17:45:39 2010 -0600

    ezmlm-cgi: Fix unclosed file descriptor on error path in msg2hash

commit 9b80a9f2e3ed0ea3c6ff832db2ce2ae9f9ceb988
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Fri Nov 12 11:46:31 2010 -0600

    unfoldHDR: Fix missing pointer increment and var initialization
    
    The unfoldHDR function was missing a pointer increment when handling
    iso-2022-{cn,kr} character sets, preventing them from being used
    properly.  As far as I can tell, this bug was introduced prior to
    version 0.42.
    
    Also, the cset variable was potentially used unitialized in the same
    function.  This problem was only revealed when the unreachable code
    caused by the above bug was corrected.

commit 86c10c0e809157a763bffa5e9b1758a8af000869
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Fri Nov 12 10:15:12 2010 -0600

    ezmlm-cron: Fix space parsing bug
    
    Comparing
    	(line.s[pos] == ' ' && line.s[pos] == '\t')
    is always false, should read "||" instead.

commit b577c7c7d952f2c698eb3ae8d3ce92f876ffe222
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Fri Nov 12 09:50:31 2010 -0600

    getln2: Fix signed/unsigned comparison

commit 59afcf01d0562f8191636f753aae5b70ffce54d9
Author: Sami Farin <sami.farin@gmail.com>
Date:   Fri Nov 12 09:49:46 2010 -0600

    getln2: Fix unreachable code preventing error propagation
    
    Date: Fri, 12 Nov 2010 17:26:46 +0200
    From: Sami Farin <sami.farin@gmail.com>
    To: Bruce Guenter <bruce@untroubled.org>
    Subject: ezmlm-idx-7.1.0 getln2 bug
    Message-ID: <20101112152646.p2yceuvxa3kxrg76@m.safari.iki.fi>
    
    this is never true:
        if (n < 0) return -1;
    therefore errors in read/write/etc are never reported to callers
    and in case of errors value of 4294967295 is being treated as legit for `n'..

commit 0f2acf8eea38932d3576ab54c6167a2270a23f94
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Fri Nov 12 09:43:54 2010 -0600

    conf-cc: Add -Wunreachable-code
    
    This adds warnings for code that cannot be reached.  For example,
    testing if unsigned variables are less than zero is always false.

commit 3aaf53d2afc6140ba0ff1b05fe47ef19cce49938
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Fri Nov 12 09:32:44 2010 -0600

    Bump version to 7.1.1

commit 710f35bda86ba6ed2154188e3b591f5d826b33a4
Author: Tullio Andreatta <t.andreatta@troppoavanti.it>
Date:   Thu Nov 11 20:32:58 2010 -0600

    lang/it: Updated messages and 3 texts

commit 639c4f63cbfecab3b12b419190421621f39868d2
Author: Lars Kruse <devel@sumpfralle.de>
Date:   Mon Nov 8 09:17:58 2010 -0600

    lang/de: fixed typo "Ergebniss" -> "Ergebnis" in messages

commit 027863d0723af474b47f859c2af14e85ec53c7b2
Author: Johannes Weberhofer <jweberhofer@weberhofer.at>
Date:   Thu Nov 4 07:28:22 2010 -0600

    Added missing items in the German messages text

commit 5629776126fad669038bc9c7bff089f59da7e98c
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Wed Nov 3 19:25:01 2010 -0600

    Tag release with today's date and reorder the changes file

commit 25484f4057c6a69526d148bb29884176e3a48088
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Wed Nov 3 17:36:29 2010 -0600

    Add note about allow/deny lists and wildcards to ezmlm.5 man page

commit 5e3fe068c9ef9a38cdc6d8b62f853370d5e9b472
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Wed Nov 3 17:27:58 2010 -0600

    Fix formatting of ezmlm-[un]sub examples in ezmlm.5 man page

commit 6309a79684c50414762a18c60f3a33ffef3258d2
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Tue Nov 2 09:53:42 2010 -0600

    Use die_sender instead of strerr_die2x(100,FATAL,MSG(ERR_NOSENDER))

commit d71e9117f1491770e630682185f18d2bb753c6c3
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Wed Oct 27 11:11:45 2010 -0600

    Add note about "omitbottom" control file to ezmlm(5) man page

commit 4109182271230144ddb23d01e7ce69567c96f413
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Wed Oct 27 10:24:04 2010 -0600

    Added tests for omitbottom control file

commit 01aeb156031dacd57e219fd5b1136630dee9e638
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Wed Oct 27 10:21:17 2010 -0600

    Fix typo in ezmlm-request man page

commit d1b04a70646c39e538750727459de20c9537adfc
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Wed Oct 27 09:54:28 2010 -0600

    ezmlm-get: Fix to use the "omitbottom" control as documented

commit c5440c4f1024efcdf78a431a35cf650f92654e66
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Tue Oct 26 16:51:42 2010 -0600


qmail/idedit.c
#include <sys/types.h>
#include <pwd.h>
#include <grp.h>
#include "readwrite.h"
#include "exit.h"
#include "scan.h"
#include "fmt.h"
#include "strerr.h"
#include "open.h"
#include "seek.h"
#include "fork.h"

#define FATAL "idedit: fatal: "
#define WARNING "idedit: warning: "

int fd;

void byte(pos,value)
char *pos;
unsigned int value;
{
  unsigned long u;
  unsigned char ch;

  if (pos[scan_ulong(pos,&u)]) return;

  if (seek_set(fd,(seek_pos) u) == -1)
    strerr_die2sys(111,FATAL,"unable to seek: ");

  ch = value;
  if (write(fd,&ch,1) != 1)
    strerr_die2sys(111,FATAL,"unable to write: ");
}

char *args[10];

void run()
{
  int pid;
  int wstat;

  pid = fork();
  if (pid == -1)
    strerr_die2sys(111,FATAL,"unable to fork: ");

  if (pid == 0) {
    execv(*args,args);
    strerr_die4sys(111,WARNING,"unable to run ",*args,": ");
  }

  if (wait_pid(&wstat,pid) != pid)
    strerr_die2sys(111,FATAL,"waitpid surprise");
}

void u(account,group,home,pos0,pos1,pos2,pos3)
char *account;
char *group;
char *home;
char *pos0;
char *pos1;
char *pos2;
char *pos3;
{
  struct passwd *pw;
  unsigned int value;

  pw = getpwnam(account);

  if (!pw && group) {
    args[0] = "add-account";
    args[1] = account;
    args[2] = group;
    args[3] = home;
    args[4] = 0;
    run();
    pw = getpwnam(account);
  }

  if (!pw)
    strerr_die3x(111,FATAL,"unable to find uid for ",account);

  value = pw->pw_uid;
  byte(pos0,value); value >>= 8;
  byte(pos1,value); value >>= 8;
  byte(pos2,value); value >>= 8;
  byte(pos3,value); value >>= 8;
  if (value)
    strerr_die3x(111,FATAL,"excessively large uid for ",account);
}

void g(group,pos0,pos1,pos2,pos3)
char *group;
char *pos0;
char *pos1;
char *pos2;
char *pos3;
{
  struct group *gr;
  unsigned int value;

  gr = getgrnam(group);

  if (!gr) {
    args[0] = "add-group";
    args[1] = group;
    args[2] = 0;
    run();
    gr = getgrnam(group);
  }

  if (!gr)
    strerr_die3x(111,FATAL,"unable to find gid for ",group);

  value = gr->gr_gid;
  byte(pos0,value); value >>= 8;
  byte(pos1,value); value >>= 8;
  byte(pos2,value); value >>= 8;
  byte(pos3,value); value >>= 8;
  if (value)
    strerr_die3x(111,FATAL,"excessively large gid for ",group);
}

int main(argc,argv)
int argc;
char **argv;
{
  if (argc < 42) _exit(100);

  fd = open_write(argv[1]);
  if (fd == -1) strerr_die4sys(111,FATAL,"unable to open ",argv[1],": ");

  g("qmail",argv[34],argv[35],argv[36],argv[37]);
  g("nofiles",argv[38],argv[39],argv[40],argv[41]);

  u("root",(char *) 0,"/",argv[14],argv[15],argv[16],argv[17]);

  u("qmaild","nofiles","/var/qmail",argv[6],argv[7],argv[8],argv[9]);
  u("qmaill","nofiles","/var/qmail",argv[10],argv[11],argv[12],argv[13]);
  u("qmailp","nofiles","/var/qmail",argv[18],argv[19],argv[20],argv[21]);
  u("alias","nofiles","/var/qmail/alias",argv[2],argv[3],argv[4],argv[5]);

  u("qmailq","qmail","/var/qmail",argv[22],argv[23],argv[24],argv[25]);
  u("qmailr","qmail","/var/qmail",argv[26],argv[27],argv[28],argv[29]);
  u("qmails","qmail","/var/qmail",argv[30],argv[31],argv[32],argv[33]);

  _exit(0);
}
